core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala

```diff
@@ -460,25 +493,213 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
+                '<br>Shuffle Write Time: ${UIUtils.formatDuration(shuffleWriteTime)}' +
+                '<br>Result Serialization Time: ${UIUtils.formatDuration(serializationTime)}' +
+                '<br>Getting Result Time: ${UIUtils.formatDuration(gettingResultTime)}">' +
+           |    '<svg class="task-assignment-timeline-duration-bar">' +
+           |    '<rect class="scheduler-delay-proportion" ' +
+           |      'x="${schedulerDelayProportionPos}%" y="0px" height="26px"' +
+           |      'width="${schedulerDelayProportion}%""></rect>' +
+           |    '<rect class="deserialization-time-proportion" '+
+           |      'x="${deserializationTimeProportionPos}%" y="0px" height="26px"' +
+           |      'width="${deserializationTimeProportion}%"></rect>' +
+           |    '<rect class="shuffle-read-time-proportion" ' +
+           |      'x="${shuffleReadTimeProportionPos}%" y="0px" height="26px"' +
+           |      'width="${shuffleReadTimeProportion}%"></rect>' +
+           |    '<rect class="executor-runtime-proportion" ' +
+           |      'x="${executorRuntimeProportionPos}%" y="0px" height="26px"' +
+           |      'width="${executorComputingTimeProportion}%"></rect>' +
+           |    '<rect class="shuffle-write-time-proportion" ' +
+           |      'x="${shuffleWriteTimeProportionPos}%" y="0px" height="26px"' +
+           |      'width="${shuffleWriteTimeProportion}%"></rect>' +
+           |    '<rect class="serialization-time-proportion" ' +
+           |      'x="${serializationTimeProportionPos}%" y="0px" height="26px"' +
+           |      'width="${serializationTimeProportion}%"></rect>' +
+           |    '<rect class="getting-result-time-proportion" ' +
+           |      'x="${gettingResultTimeProportionPos}%" y="0px" height="26px"' +
+           |      'width="${gettingResultTimeProportion}%"></rect></svg>',
+           |  'start': new Date(${launchTime}),
+           |  'end': new Date(${finishTime})
+           |}
+         """.stripMargin
+      timelineObject
+    }.mkString("[", ",", "]")
+
+    val groupArrayStr = executorsSet.map {
+      case (executorId, host) =>
+        s"""
+            |{
+            |  'id': '${executorId}',
+            |  'content': '${executorId} / ${host}',
+            |}
+          """.stripMargin
+    }.mkString("[", ",", "]")
+
+    var maxWindowInSec = ((maxFinishTime - minLaunchTime) / 1000.0).round
+    if (maxWindowInSec <= 0) maxWindowInSec = 1
+    val tasksPerSecond = numEffectiveTasks / maxWindowInSec
+    var maxZoom = {
+      if (tasksPerSecond > 100) {
+        1000L / (tasksPerSecond / 100)
+      }
+      else {
+        1000 * (100 / tasksPerSecond)
+      }
+    }
+
+    if (maxZoom < 0) maxZoom = 1
+
+    <span class="expand-task-assignment-timeline">
+      <span class="expand-task-assignment-timeline-arrow arrow-closed"></span>
+      <a>Event Timeline (Most recent {numEffectiveTasks.min(MAX_TIMELINE_TASKS)} tasks)</a>
```

Comment:
Only say "Most recent X" if we have truncated the list. If there are less than MAX_TIMELINE_TASKS overall, I'd just say "Event Timeline".